### ISSUE TITLE
nanoseconds timestamps  display wrong with -tt option

### ISSUE THREAD

**Issue by Vasikby** [the-tcpdump-group/tcpdump#466]
[vasya@foo /]$ tcpdump --help
tcpdump version 4.7.4
libpcap version 1.7.3
OpenSSL 1.0.1k-fips 8 Jan 2015

I captured some packets with command: 
[vasya@foo /]$ sudo tcpdump -w /tmp/eno1.pcap -n --time-stamp-precision=nano -i eno1

I dumped timestamps with -tt and without -tt option:
[vasya@foo /]$ tcpdump -tt -n -r /tmp/eno1.pcap  --time-stamp-precision=nano | awk '{print $1}'
... last 5 packets below ...
...
1434463413.859717812
1434463414.609823284
1434463415.94652557    - wrong
1434463415.359737677
1434463417.111141442

[vasya@foo /]$ tcpdump  -n -r /tmp/eno1.pcap  --time-stamp-precision=nano | awk '{print $1}'
... last 5 packets below ...
...
17:03:33.859717812
17:03:34.609823284
17:03:35.094652557   - right
17:03:35.359737677
17:03:37.111141442

As result timestamps for nanoseconds are wrong for some packets (it occur with leading zero)

**Comment by fxlb** [comment]
Thanks for the report.
Could you try this patch?
```patch
diff --git a/util-print.c b/util-print.c
index a46e12c..b52d277 100644
--- a/util-print.c
+++ b/util-print.c
@@ -213,7 +213,7 @@ ts_print(netdissect_options *ndo,
                break;
 
        case 2: /* Unix timeval style */
-               ND_PRINT((ndo, "%u.%06u ",
+               ND_PRINT((ndo, "%u.%09u ",
                             (unsigned)tvp->tv_sec,
                             (unsigned)tvp->tv_usec));
                break;
```

**Comment by fxlb** [comment]
This need a better patch. Working on.

**Comment by fxlb** [comment]
Thanks for the report.
Could you try this patch?
```patch
diff --git a/util-print.c b/util-print.c
index a46e12c..f43af8b 100644
--- a/util-print.c
+++ b/util-print.c
@@ -188,6 +188,43 @@ _U_
 }
 
 /*
+ * Format the timestamp - Unix timeval style
+ */
+static char *
+ts_unix_format(netdissect_options *ndo
+#ifndef HAVE_PCAP_SET_TSTAMP_PRECISION
+_U_
+#endif
+, int sec, int usec)
+{
+	static char buf[sizeof("0000000000.000000000")];
+	const char *format;
+
+#ifdef HAVE_PCAP_SET_TSTAMP_PRECISION
+	switch (ndo->ndo_tstamp_precision) {
+
+	case PCAP_TSTAMP_PRECISION_MICRO:
+		format = "%u.%06u";
+		break;
+
+	case PCAP_TSTAMP_PRECISION_NANO:
+		format = "%u.%09u";
+		break;
+
+	default:
+		format = "%u.{unknown precision}";
+		break;
+	}
+#else
+	format = "%u.%06u";
+#endif
+
+	snprintf(buf, sizeof(buf), format, (unsigned)sec, (unsigned)usec);
+
+	return buf;
+}
+
+/*
  * Print the timestamp
  */
 void
@@ -213,9 +250,8 @@ ts_print(netdissect_options *ndo,
 		break;
 
 	case 2: /* Unix timeval style */
-		ND_PRINT((ndo, "%u.%06u ",
-			     (unsigned)tvp->tv_sec,
-			     (unsigned)tvp->tv_usec));
+		ND_PRINT((ndo, "%s ", ts_unix_format(ndo, tvp->tv_sec,
+			  tvp->tv_usec)));
 		break;
 
 	case 3: /* Microseconds since previous packet */
```

**Comment by gvanem** [comment]
Where there not plans to use the dissectors and printing routines in a library? If so, that:
```
 static char buf[sizeof("0000000000.000000000")];
```
isn't reentrant AFAICS. But you could use a local buf to pass to `ts_unix_format()`.

**Comment by fxlb** [comment]
You mean like this?
```patch
diff --git a/util-print.c b/util-print.c
index a46e12c..856d28e 100644
--- a/util-print.c
+++ b/util-print.c
@@ -56,6 +56,8 @@
 
 int32_t thiszone;		/* seconds offset from gmt to local time */
 
+#define MAX_SIZE_TS_UNIX_FORMAT sizeof("0000000000.000000000")
+
 /*
  * Print out a null-terminated filename (or other ascii string).
  * If ep is NULL, assume no truncation check is needed.
@@ -188,6 +190,43 @@ _U_
 }
 
 /*
+ * Format the timestamp - Unix timeval style
+ */
+static char *
+ts_unix_format(netdissect_options *ndo
+#ifndef HAVE_PCAP_SET_TSTAMP_PRECISION
+_U_
+#endif
+, int sec, int usec, char *buf)
+{
+	const char *format;
+
+#ifdef HAVE_PCAP_SET_TSTAMP_PRECISION
+	switch (ndo->ndo_tstamp_precision) {
+
+	case PCAP_TSTAMP_PRECISION_MICRO:
+		format = "%u.%06u";
+		break;
+
+	case PCAP_TSTAMP_PRECISION_NANO:
+		format = "%u.%09u";
+		break;
+
+	default:
+		format = "%u.{unknown precision}";
+		break;
+	}
+#else
+	format = "%u.%06u";
+#endif
+
+	snprintf(buf, MAX_SIZE_TS_UNIX_FORMAT, format,
+		 (unsigned)sec, (unsigned)usec);
+
+	return buf;
+}
+
+/*
  * Print the timestamp
  */
 void
@@ -213,9 +252,11 @@ ts_print(netdissect_options *ndo,
 		break;
 
 	case 2: /* Unix timeval style */
-		ND_PRINT((ndo, "%u.%06u ",
-			     (unsigned)tvp->tv_sec,
-			     (unsigned)tvp->tv_usec));
+		{
+			char buf[MAX_SIZE_TS_UNIX_FORMAT];
+			ND_PRINT((ndo, "%s ", ts_unix_format(ndo,
+				  tvp->tv_sec, tvp->tv_usec, buf)));
+		}
 		break;
 
 	case 3: /* Microseconds since previous packet */
```

**Comment by fxlb** [comment]
I will do a similar update in `ts_format`.

**Comment by gvanem** [comment]
@fxlb Much better.

**Comment by fxlb** [comment]
Fixed by 22d4afb88abd348cae3097639585e0f6fd5ddc95.
